flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java

```diff
@@ -1256,49 +1307,82 @@ public void testProcessWithUnAvailableOutput() throws Exception {
             executor.submit(
-                    () -> {
-                        executor.submit(
-                                completeFutureTask,
-                                "This task will complete the future to resume process input action.");
-                    },
+                    () ->
+                            new WaitingThread(
+                                            executor,
+                                            completeFutureTask,
+                                            sleepTimeInsideMail,
+                                            sleepTimeOutsideMail)
+                                    .start(),
                     "This task will submit another task to execute after processing input once.");
 
+            long startTs = System.currentTimeMillis();
             TaskIOMetricGroup ioMetricGroup =
                     task.getEnvironment().getMetricGroup().getIOMetricGroup();
             task.invoke();
+            long totalDuration = System.currentTimeMillis() - startTs;
             assertThat(
                     ioMetricGroup.getBackPressuredTimePerSecond().getCount(),
-                    Matchers.greaterThanOrEqualTo(sleepTime));
+                    Matchers.greaterThanOrEqualTo(sleepTimeOutsideMail));
+            assertThat(
+                    ioMetricGroup.getBackPressuredTimePerSecond().getCount(),
+                    Matchers.lessThanOrEqualTo(totalDuration - sleepTimeInsideMail));
             assertThat(ioMetricGroup.getIdleTimeMsPerSecond().getCount(), is(0L));
             assertEquals(numberOfProcessCalls, inputProcessor.currentNumProcessCalls);
         }
     }
 
     @Test
     public void testProcessWithUnAvailableInput() throws Exception {
-        final long unAvailableTime = 42;
+        final long sleepTimeOutsideMail = 42;
+        final long sleepTimeInsideMail = 44;
         try (final MockEnvironment environment = setupEnvironment(new boolean[] {true, true})) {
             final UnAvailableTestInputProcessor inputProcessor =
-                    new UnAvailableTestInputProcessor(unAvailableTime);
+                    new UnAvailableTestInputProcessor();
             final StreamTask task =
                     new MockStreamTaskBuilder(environment)
                             .setStreamInputProcessor(inputProcessor)
                             .build();
 
+            final MailboxExecutor executor = task.mailboxProcessor.getMainMailboxExecutor();
+            final RunnableWithException completeFutureTask =
+                    () -> {
+                        inputProcessor
+                                .availabilityProvider
+                                .getUnavailableToResetAvailable()
+                                .complete(null);
+                    };
+
+            // Make sure WaitingThread is started after Task starts processing.
+            executor.submit(
+                    () ->
+                            new WaitingThread(
```

Comment:
Why do we need a thread instead of directly executing the content of the mail? Is it because we wouldn't be able to measure any idle time?

Reply:
Yes. We need to sleep outside mailbox action for the time being accounted as idle. (That's the point of this bug fix)

Reply:
I was thinking if it's safer to also join the `WaitingThread` after `task.invoke();` but I guess that's more or less done through the `completeFutureTask` (should that be called in a `finally`?).